src/parser/validator.cc

```diff
@@ -0,0 +1,218 @@
+#include "maliput_sparse/parser/validator.h"
+
+#include "maliput_sparse/geometry/utility/geometry.h"
+#include "maliput_sparse/parser/junction.h"
+#include "maliput_sparse/parser/lane.h"
+#include "maliput_sparse/parser/segment.h"
+
+namespace maliput_sparse {
+namespace parser {
+
+namespace {
+
+// Convenient constant for indicating left or right.
+static constexpr bool kLeft{true};
+static constexpr bool kRight{!kLeft};
+
+}  // namespace
+
+Validator::Validator(const Parser* parser, const ValidatorOptions& options, const ValidatorConfig config) {
+  MALIPUT_THROW_UNLESS(parser);
+  if (options.lane_adjacency) {
+    ValidateLaneAdjacency(parser, config);
+  }
+}
+
+const std::vector<Validator::Error>& Validator::GetErrors() const { return errors_; }
+
+void Validator::CheckAdjacency(const Lane& lane, const Lane& adjacent_lane, bool left, double tolerance) {
+  const auto line_string_a = left ? lane.left : lane.right;
+  const auto line_string_b = left ? adjacent_lane.right : adjacent_lane.left;
+  const double distance = geometry::utility::ComputeDistance(line_string_a, line_string_b);
+  if (distance > tolerance) {
+    const std::string msg{"Lane " + lane.id + " and lane " + adjacent_lane.id +
+                          " are not adjacent under the tolerance " + std::to_string(tolerance) +
+                          ". (distance: " + std::to_string(distance) + ")"};
+    Report(msg, Validator::Error::Type::kLaneAdjacency, Validator::Error::Severity::kError);
+  }
+}
+
+void Validator::Report(const std::string& message, const Validator::Error::Type& type,
+                       const Validator::Error::Severity& severity) {
+  errors_.push_back(Validator::Error{message, type, severity});
+}
+
+void Validator::ValidateLaneAdjacency(const Parser* parser, const ValidatorConfig config) {
+  const auto junctions = parser->GetJunctions();
+  for (const auto junction : junctions) {
+    for (const auto segment : junction.second.segments) {
+      const std::vector<Lane>& lanes = segment.second.lanes;
+      // Iterates lanes from right to left.
+      // | idx=N-1 | idx=3 | idx=2 | idx=1 | idx=0 |
+      for (int idx{}; idx < static_cast<int>(lanes.size()); ++idx) {
+        const Lane& lane = lanes[idx];
+
+        // Note: The following code is a bit repetitive for left and right adjacency checks, but it is easier to read
+        // and understand. Check right adjacency <--------------------------------------------------
+        if (lane.right_lane_id) {
+          // Check if there is a idx to the right.
+          if (idx == 0) {
+            // Error.
+            const std::string msg{"Wrong ordering of lanes: Lane " + lane.id + " has a right lane id (" +
+                                  lane.right_lane_id.value() + ") but is the first lane in the segment " +
+                                  segment.first + "."};
+            Report(msg, Validator::Error::Type::kLaneAdjacency, Validator::Error::Severity::kError);
+          }
```

Comment:
I think you can significantly reduce the number of lines of code by doing something like the following:

```C++
std::string<Error> errors;
auto evaluate = [errors](bool condition, const std::string& message) {
  if (condition) {
    errors.emplace_back(message, Validator::Error::Type::kLaneAdjacency, Validator::Error::Severity::kError);
  }
 return condition;
};
```

And then use it:

```c++
const std::string msg{"Wrong ordering of lanes: Lane " + lane.id + " has a right lane id (" +
    lane.right_lane_id.value() + ") but is the first lane in the segment " +
    segment.first + "."};
evaluate(idx == 0, msg);
```

Reply:
Done